### ISSUE TITLE
Utilizing Automatic (algorithmic) differentiation

### ISSUE THREAD

**Issue by michaelbynum** [OpenWaterAnalytics/next-net#4]
I would like to propose the possible use of automatic differentiation (AD) for computing derivatives within Epanet. 

What is automatic differentiation?
----------------------------------
Automatic differentiation is a way to compute derivatives accurately and efficiently from a mathematical expression or function. It is neither finite difference nor symbolic differentiation. Automatic differentiation employs the chain rule on elementary operations (+, -, *, /, log, cos, etc.) to also compute derivatives as a function (or expression) is evaluated (at the current values of the variables). There is a great FAQ at [http://www.autodiff.org/?module=Introduction&submenu=FAQ](http://www.autodiff.org/?module=Introduction&submenu=FAQ) that gives some basic ideas. A couple papers:

- Griewank, Andreas, David Juedes, and Jean Utke. "Algorithm 755: ADOL-C: a package for the automatic differentiation of algorithms written in C/C++." ACM Transactions on Mathematical Software (TOMS) 22.2 (1996): 131-167. [https://dl.acm.org/citation.cfm?id=229474](https://dl.acm.org/citation.cfm?id=229474)

- Andersson, Joel, Johan Åkesson, and Moritz Diehl. "CasADi: A symbolic package for automatic differentiation and optimal control." Recent advances in algorithmic differentiation. Springer, Berlin, Heidelberg, 2012. 297-307. [https://link.springer.com/chapter/10.1007%2F978-3-642-30023-3_27](https://link.springer.com/chapter/10.1007%2F978-3-642-30023-3_27)

How can AD be used in Epanet?
--------------------------------
One possible way of solving a set of nonlinear equations, C(x) = 0 is with Newton's method. The core of this algorithm involves solving a set of linear equations, Ax = b, where A is the jacobian of C(x) at the current estimate of x. Note that 2 of the 3 hydraulic solvers that Lew Rossman proposed (#2) include Newton iterations. 

For every constraint/equation (head loss model, demand model, mass balance, etc.), Epanet needs a way to compute both the residual of the constraint and the derivative of the constraint with respect to each variable in the constraint. If AD is used for the derivatives, only the evaluation of the residual of the constraint needs to be implemented. This would make Epanet far more easily extensible. For example, if a new model is needed, only the residual of the constraint needs to be implemented. This increases productivity and is less error prone.

What tools exist to perform AD?
--------------------------------
There are several existing tools that do automatic differentiation. A few are:

- CasADi: [https://github.com/casadi/casadi/wiki](https://github.com/casadi/casadi/wiki)

- ADOL-C: [https://projects.coin-or.org/ADOL-C](https://projects.coin-or.org/ADOL-C)

- CppAD: [https://www.coin-or.org/CppAD/](https://www.coin-or.org/CppAD/)

- A more complete list can be found at [http://www.autodiff.org/?module=Tools](http://www.autodiff.org/?module=Tools)

**Comment by samhatchett** [comment]
This is really fascinating reading. Thanks! I've tagged this issue in a way I thought appropriate, but if you can think of a better taxonomy let me know.

**Comment by LRossman** [comment]
I agree with @samhatchett that this AD stuff looks cool but I wonder if it would really buy one anything for solving network hydraulics. The nonlinear equations only involve head loss (h) as a function of flow (q) which for the Hazen-Williams model is: h = r q^n where r and n are constants. The derivative of this is simply  nh/q so what's to be gained by using AD?

On a related note, one of the links on the pages listed by @michaelbynum was to the [fREEDA Project](https://www.freeda.org/) which is an electronic circuit simulator in the spirit of SPICE (SPICE is to electronic circuits as EPANET is to hydraulic networks, only more extensive). The [Programmers Manual](https://www.freeda.org/doc/programmers_manual/freeda_programmers.pdf) describes some architectural features of fREEDA that might be applicable to next-net.

**Comment by jamesuber** [comment]
I think the use case for AD is not computing the jacobian of the basic hydraulic equations. The use case should be things like "what's the derivative of pressure at node N and time T with respect to a change in demand at node Y?" This is a problem that, to my knowledge, has never even been provided a solution in the literature. People compute it just using finite differences, which presents problems for solving problems like parameter estimation (both for efficiency and accuracy). 

I'm not necessarily strongly advocating for AD -- but primarily cause I've never used it and therefor don't know much about it. But being able to efficiently and accurately compute the above sorts of derivatives definitely feeds into a variety of use cases
